waiter/src/waiter/handler.clj

```diff
@@ -254,34 +254,57 @@
+  (let [service-description-params (select-keys request-params sd/service-parameter-keys)
+        param-predicates (map (fn [[param-name param-value]]
+                                (let [param-predicate (param-value->filter-fn param-value)]
+                                  (fn [service-description]
+                                    (param-predicate (str (get service-description param-name))))))
+                              (seq service-description-params))]
+    (fn [service-description]
+      (or (empty? param-predicates)
+          (every? #(%1 service-description) param-predicates)))))
+
 (defn list-services-handler
   "Retrieves the list of services viewable by the currently logged in user.
    A service is viewable by the run-as-user or a waiter super-user."
   [entitlement-manager query-state-fn query-autoscaler-state-fn prepend-waiter-url service-id->service-description-fn
    service-id->metrics-fn service-id->references-fn service-id->source-tokens-entries-fn request]
   (let [{:keys [all-available-service-ids service-id->healthy-instances service-id->unhealthy-instances] :as global-state} (query-state-fn)]
     (let [{:strs [run-as-user token token-version] :as request-params} (-> request ru/query-params-request :query-params)
+          service-description-filter-predicate (query-params->service-description-filter-predicate
+                                                 (dissoc request-params "run-as-users"))
           auth-user (get request :authorization/user)
-          viewable-services (filter
-                              (fn [service-id]
-                                (let [service-description (service-id->service-description-fn service-id :effective? false)
-                                      source-tokens (-> (service-id->source-tokens-entries-fn service-id)
-                                                        vec
-                                                        flatten)]
-                                  (and (if (str/blank? run-as-user)
-                                         (authz/manage-service? entitlement-manager auth-user service-id service-description)
-                                         ((str->filter-fn run-as-user) (get service-description "run-as-user")))
-                                       (or (str/blank? token)
-                                           (let [filter-fn (str->filter-fn token)]
-                                             (->> source-tokens
+          viewable-service-ids (filter
+                                 (fn [service-id]
+                                   (let [service-description (service-id->service-description-fn service-id :effective? true)
+                                         source-tokens (-> (service-id->source-tokens-entries-fn service-id) vec flatten)]
+                                     (and (service-description-filter-predicate service-description)
+                                          (if (or (nil? run-as-user)
```

Comment:
I assume this nil? guard was to protect against a bug? Why only here, not token-version? or token?

Reply:
The `nil` behavior for `run-as-user` is different for legacy reasons. When no `run-as-user` is provided, we display only services a user can manage.